src/omv/ports/mimxrt/omv_gpu.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * This file is part of the OpenMV project.
+ *
+ * Copyright (c) 2013-2024 Ibrahim Abdelkader <iabdalkader@openmv.io>
+ * Copyright (c) 2013-2024 Kwabena W. Agyeman <kwagyeman@openmv.io>
+ *
+ * This work is licensed under the MIT license, see the file LICENSE for details.
+ *
+ * GPU driver for IMXRT port.
+ */
+#include "omv_boardconfig.h"
+#if (OMV_GPU_ENABLE == 1)
+#include "imlib.h"
+
+#include "fsl_pxp.h"
+
+#define TIME_GPU    (0)
+#if (TIME_GPU == 1)
+#include "py/mphal.h"
+#include <stdio.h>
+#endif
+
+int omv_gpu_init() {
+    return 0;
+}
+
+void omv_gpu_deinit() {
+    PXP_Deinit(PXP);
+}
+
+int omv_gpu_draw_image(image_t *src_img,
+                       rectangle_t *src_rect,
+                       image_t *dst_img,
+                       rectangle_t *dst_rect,
+                       int alpha,
+                       const uint16_t *color_palette,
+                       const uint8_t *alpha_palette,
+                       image_hint_t hint) {
+    // PXP input image must be GRAYSCALE, RGB565, YUV422 or YVU422.
+    if ((src_img->pixfmt != PIXFORMAT_GRAYSCALE) && (src_img->pixfmt != PIXFORMAT_RGB565) &&
+        (src_img->pixfmt != PIXFORMAT_YUV422) && (src_img->pixfmt != PIXFORMAT_YVU422)) {
+        return -1;
+    }
+
+    // PXP output image must be GRAYSCALE or RGB565.
+    if ((dst_img->pixfmt != PIXFORMAT_GRAYSCALE) && (dst_img->pixfmt != PIXFORMAT_RGB565)) {
+        return -1;
+    }
+
+    // If the output image is GRAYSCALE, the input image must be GRAYSCALE or YUV422.
+    if ((dst_img->pixfmt == PIXFORMAT_GRAYSCALE) && (src_img->pixfmt != PIXFORMAT_GRAYSCALE) &&
+        (src_img->pixfmt != PIXFORMAT_YUV422) && (src_img->pixfmt != PIXFORMAT_YVU422)) {
+        return -1;
+    }
+
+    // PXP can only operate on 8x8 or 16x16 pixel blocks.
+    if ((src_rect->w % 8) || (src_rect->h % 8) || (dst_rect->w % 8) || (dst_rect->h % 8)) {
+        return -1;
+    }
+
+    // PXP cannot alpha blend on the background in-place and doesn't have a LUT for alpha or color palettes.
+    if ((alpha != 256) || (color_palette || alpha_palette)) {
+        return -1;
+    }
+
+    // PXP cannot reduce the image size by more than 16x.
+    if ((dst_rect->w > (src_rect->w * 16)) || (dst_rect->h > (src_rect->h * 16))) {
+        return -1;
+    }
+
+    // PXP cannot scale the image size by more than 4096x.
+    if ((dst_rect->w < (src_rect->w / 4096)) || (dst_rect->h < (src_rect->h / 4096))) {
+        return -1;
+    }
+
+    // PXP cannot hmirror or vflip when scaling.
+    if ((hint & (IMAGE_HINT_HMIRROR | IMAGE_HINT_VFLIP)) &&
+        ((src_rect->w != dst_rect->w) || (src_rect->h != dst_rect->h))) {
+        return -1;
+    }
+
+    // PXP always applies bilinear scaling, which is always better than nearest-neighbor so ignore the hint.
+
+    #if (TIME_GPU == 1)
+    mp_uint_t start = mp_hal_ticks_ms();
+    #endif
```

Comment:
We have profiling macros now, see omv_gpu in stm32. The macro should be at the start & end of the function.

Reply:
Sure